File: old-projects/src/Game.ts
import { Point } from "./Point";
import { Sprite } from "./Sprite";
import { Util } from "./Util";
import { Monster } from "./Monsters/Monster";
import { BallMonster } from "./Monsters/BallMonster";
import { SquareMonster } from "./Monsters/SquareMonster";
import { TriangleMonster } from "./Monsters/TriangleMonster";
import { ExplosionParticle } from "./Particles/ExplosionParticle";
import { Tower } from "./Towers/Tower";

export class Game {
  static readonly ticksPerSecond = 10000000;
  static readonly baseTicksPerFrame = Game.ticksPerSecond / 60;
  static readonly sprites: Sprite[] = [];
  static readonly towers: Tower[] = [];

  static monstersAllowedToEscape = 0;
  static canvasHeight = 450;
  static canvasWidth = 700;
  static currentFrameMonsters: Monster[] = [];
  static currentFrameMultiplier = 1;
  static ticksNow = 0;
  static monstersComing = false;
  static lastMonsterGone = false;
  static isPlaying = false;

  private static monstersKilled = 0;
  private static monstersEscaped = 0;
  private static money = 0;
  private static selectedTowerValue: Tower | null = null;

  static towerAnimation: Array<() => void> = [];
  static towerSelected: Array<(tower: Tower, towerWasSelectedBefore: boolean) => void> = [];
  static towerUnselected: Array<() => void> = [];
  static balanceChanged: Array<() => void> = [];
  static paused: Array<() => void> = [];
  static resumed: Array<() => void> = [];
  static monsterKilled: Array<(monster: Monster) => void> = [];
  static monsterEscaped: Array<(monster: Monster) => void> = [];

  static get selectedTower(): Tower | null {
    return Game.selectedTowerValue;
  }

  static set selectedTower(value: Tower | null) {
    Game.setSelectedTower(value);
  }

  static get balance(): number {
    return Game.money;
  }

  static resumeGame(): void {
    Game.isPlaying = true;
    Game.resumed.forEach((handler) => handler());
  }

  static pauseGame(): void {
    Game.isPlaying = false;
    Game.paused.forEach((handler) => handler());
  }

  static executeTowerAnimations(): void {
    const handlers = [...Game.towerAnimation];
    handlers.forEach((handler) => handler());
  }

  static setSelectedTower(value: Tower | null): void {
    const oldValue = Game.selectedTowerValue;
    Game.selectedTowerValue = value;
    if (Game.selectedTowerValue !== null) {
      Game.towerSelected.forEach((handler) => handler(Game.selectedTowerValue as Tower, oldValue !== null));
    } else if (oldValue !== null) {
      Game.towerUnselected.forEach((handler) => handler());
    }
  }

  static killMonster(monster: Monster): void {
    let interval = Util.calculateIntervalToComplete(0.4);
    for (let i = 0; i < 35; i += 1) {
      new ExplosionParticle(monster.x, monster.y, Math.floor(Util.randomInRange(2, 5)), monster.color, interval);
    }

    interval = Util.calculateIntervalToComplete(0.5);
    for (let i = 0; i < 35; i += 1) {
      new ExplosionParticle(monster.x, monster.y, Math.floor(Util.randomInRange(1, 4)), monster.color, interval);
    }

    Game.monstersKilled += 1;
    monster.markedForRemoval = true;
    Game.monsterKilled.forEach((handler) => handler(monster));
  }

  static escapeMonster(monster: Monster): void {
    const interval = Util.calculateIntervalToComplete(1);
    for (let i = 0; i < 150; i += 1) {
      new ExplosionParticle(monster.x, monster.y, Math.floor(Util.randomInRange(1, 4)), Util.createRandomColor(), interval);
    }
    Game.monstersEscaped += 1;
    monster.markedForRemoval = true;
    Game.monsterEscaped.forEach((handler) => handler(monster));
  }

  static createExplosionParticles(x: number, y: number, particleCount: number, particleSize: number, color: string, burnDown: number): void {
    for (let i = 0; i < particleCount; i += 1) {
      new ExplosionParticle(x, y, particleSize, color, burnDown);
    }
  }

  static resetBalance(amount: number): void {
    Game.money = amount;
    Game.balanceChanged.forEach((handler) => handler());
  }

  static deposit(amount: number): void {
    Game.money += amount;
    Game.balanceChanged.forEach((handler) => handler());
  }

  static withdraw(amount: number): boolean {
    if (amount > Game.money) {
      return false;
    }
    Game.money -= amount;
    Game.balanceChanged.forEach((handler) => handler());
    return true;
  }

  static clearField(): void {
    Game.monstersKilled = 0;
    Game.monstersEscaped = 0;
    Game.selectedTower = null;
    Game.towers.splice(0).forEach((tower) => tower.removeTowerFromCanvas());
    Game.sprites.splice(0).forEach((sprite) => sprite.removeFromCanvas());
  }

  static createRandomMonster(points: Point[]): void {
    const rand = Math.random();
    if (rand < 0.333) {
      Game.createMonster("b", points);
    } else if (rand < 0.666) {
      Game.createMonster("t", points);
    } else {
      Game.createMonster("s", points);
    }
  }

  static createMonster(code: string, points: Point[]): void {
    const start = points[0];
    switch (code) {
      case "b":
        new BallMonster(start.x, start.y, points);
        return;
      case "t":
        new TriangleMonster(start.x, start.y, points);
        return;
      default:
        new SquareMonster(start.x, start.y, points);
    }
  }
}

File: old-projects/vector-defence-silverlight-port/original-csharp-source/Game.cs
using System;
using System.Collections.Generic;
using System.Windows;
using System.Windows.Controls;
using System.Windows.Media;
using VectorDefenceSL.Monsters;
using VectorDefenceSL.Particles;
using VectorDefenceSL.Towers;

namespace VectorDefenceSL
{
    public static class Game
    {
        public const int TicksPerSecond = 10000000;
        public const double BaseTicksPerFrame = TicksPerSecond / 60; // (ticks in 1 second) / (base 60 fps target)

        public static readonly List<Sprite> Sprites = new List<Sprite>();
        public static readonly List<Tower> Towers = new List<Tower>();

        public static int MonstersAllowedToEscape;

        public static Canvas Canvas;
        public static UIElementCollection Children;

        public static double CanvasHeight;

        public static Monster[] CurrentFrameMonsters;
        public static double CurrentFrameMultiplier;
        public static long TicksNow;
        public static double CanvasWidth;

        private static readonly object moneyLock = new object();

        private static int monstersKilled;
        private static int monstersEscaped;
        private static int money;
        private static Tower selectedTowerValue;

        public static event Action TowerAnimation;

        public static event Action<Tower, bool> TowerSelected;

        public static event Action TowerUnselected;

        public static event Action BalanceChanged;

        public static event Action Paused;

        public static event Action Resumed;

        public static event Action<Monster> MonsterKilled;

        public static event Action<Monster> MonsterEscaped;

        public static Tower SelectedTower
        {
            get
            {
                return selectedTowerValue;
            }

            set
            {
                SetSelectedTower(value);            
            }
        }

        public static int Balance
        {
            get { return money; }
        }

        public static bool MonstersComing { get; set; }

        public static bool LastMonsterGone { get; set; }

        public static bool IsPlaying { get; private set; }

        public static void ResumeGame()
        {
            if (Resumed != null)
            {
                Resumed();
            }
            IsPlaying = true;
        }

        public static void PauseGame()
        {
            if (Paused != null)
            {
                Paused();
            }
            IsPlaying = false;
        }

        public static void ExecuteTowerAnimations()
        {
            if (TowerAnimation != null)
            {
                TowerAnimation();
            }
        }

        public static void SetSelectedTower(Tower value)
        {
            var oldValue = selectedTowerValue;
            selectedTowerValue = value;
            if (selectedTowerValue != null)
            {
                bool towerWasSelectedBefore = oldValue != null;
                TowerSelected(selectedTowerValue, towerWasSelectedBefore);
            }
            else if (oldValue != null)
            {
                TowerUnselected();
            }                    
        }

        public static void AddElement(UIElement e)
        {
            Children.Add(e);
        }

        public static void KillMonster(Monster m)
        {
            var interval = Util.CalculateIntervalToComplete(.4);
            for (int i = 0; i < 35; i++)
            {
                var size = Util.Rand.Next(2, 5);
                new ExplosionParticle(m.X, m.Y, size, m.Color, interval);
            }

            interval = Util.CalculateIntervalToComplete(.5);
            for (int i = 0; i < 35; i++)
            {
                var size = Util.Rand.Next(1, 4);
                new ExplosionParticle(m.X, m.Y, size, m.Color, interval);
            }

            monstersKilled++;
            m.MarkedForRemoval = true;
            MonsterKilled(m);
        }

        public static void EscapeMonster(Monster m)
        {
            var interval = Util.CalculateIntervalToComplete(1);
            for (int i = 0; i < 150; i++)
            {
                Color color = Util.CreateRandomColor();
                var size = Util.Rand.Next(1, 4);
                new ExplosionParticle(m.X, m.Y, size, color, interval);
            }             
            monstersEscaped++;
            m.MarkedForRemoval = true;
            MonsterEscaped(m);
        }

        public static void InitializeGame(Canvas c)
        {
            Canvas = c;
            CanvasHeight = Canvas.Height;
            CanvasWidth = Canvas.Width;
            Children = c.Children;
        }

        public static void CreateExplosionParticles(double x, double y, int particleCount, double particleSize, Color color, double burnDown)
        {
            for (int i = 0; i < particleCount; i++)
            {
                new ExplosionParticle(x, y, particleSize, color, burnDown);
            }
        }

        public static void ResetBalance(int amount)
        {
            lock (moneyLock)
            {
                money = amount;
            }
            BalanceChanged();
        }

        public static void Deposit(int amount)
        {
            lock (moneyLock)
            {
                money += amount;    
            }
            BalanceChanged();
        }

        public static bool Withdraw(int amount)
        {
            lock (moneyLock)
            {
                if (amount > money)
                {
                    return false;
                }
                money -= amount;
            }
            BalanceChanged();
            return true;
        }

        public static void ClearField()
        {
            monstersKilled = monstersEscaped = 0;
            SelectedTower = null;
            foreach (var t in Towers)
            {
                t.RemoveTowerFromCanvas();    
            }
            Towers.Clear();

            foreach (var s in Sprites)
            {
                s.RemoveFromCanvas();
            }
            Sprites.Clear();
        }

        public static void CreateRandomMonster(PointCollection points)
        {
            double rand = Util.Rand.NextDouble();
            if (rand < .333)
            {
                CreateMonster('b', points);
            }
            else if (rand >= .333 && rand < .666)
            {
                CreateMonster('t', points);
            }
            else
            {
                CreateMonster('s', points);
            }
        }

        public static void CreateMonster(char c, PointCollection points)
        {
            Point start = points[0];
            switch (c)
            {
                case 'b':
                    new BallMonster(start.X, start.Y, points);
                    return;
                case 't':
                    new TriangleMonster(start.X, start.Y, points);
                    return;
                default:
                    new SquareMonster(start.X, start.Y, points);
                    return;
            }
        }
    }
}
